frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +3985,96 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
   return isTypeDefaultInitializableQuery(context, t);
 }
 
+// First is from const, second is from ref.
+static const std::pair<bool, bool> getCopyabilityInfoImpl(
+    Context* context, const CompositeType* ct) {
+  std::pair<bool, bool> result = {false, false};
+
+  // Get initial values directly from copyability flags, if present.
+  auto ast = parsing::idToAst(context, ct->id());
+  auto attrs = ast->attributeGroup();
+  if (attrs) {
+    result.first = attrs->hasPragma(PRAGMA_TYPE_INIT_EQUAL_FROM_CONST);
+    result.second = attrs->hasPragma(PRAGMA_TYPE_INIT_EQUAL_FROM_REF);
```

Comment:
I was just randomly traversing the PR list and happened to take a look at the changes here.

I don't know if in Dyno we need to handle these attributes. I believe the reason they exist in production is to cache work: types are determined to be "copyable" recursively, so if every query on copyability re-traversed the type and its fields, it would be quite slow. So instead, once we determine that a type is copyable, we add a flag to it, so that later queries don't have to traverse it, and can just consult the flags. Thus, the flags you're checking for here are markers added by the production compiler's algorithm (I think). c.f. `setRecordCopyableFlags` in `preFold.cpp`. Since types in Dyno are immutable, we shouldn't need to worry about these flags getting set and reset.

On the other hand, the use of these flags indicates (to me) that we should be caching as much of this as possible. In other words, it would be best to make sure queries are being used, through the recursive traversal of a type. I don't actually know how resolving `init=` takes care of this -- just something to note.

Reply:
Sounds good. I was actually just porting over the code from `setRecordCopyableFlags` minus setting flags.
As far as caching, a recursive query seems like the most natural approach, but those are illegal or at least considered a bad practice right?

Reply:
It's only illegal if it depends on itself somehow. query(pair(a,b)) = query(a), query(b) is valid and fine.